state-sync/storage-service/server/src/lib.rs

```diff
@@ -166,25 +373,144 @@ impl<T: StorageReaderInterface> StorageServiceServer<T> {
-                            .call(protocol, request);
-                    log_storage_response(&response);
-                    response_sender.send(response);
+                    Handler::new(
+                        cached_storage_server_summary,
+                        data_subscriptions,
+                        lru_storage_cache,
+                        storage,
+                        time_service,
+                    )
+                    .process_request_and_respond(
+                        peer,
+                        protocol,
+                        request,
+                        response_sender,
+                    );
                 })
                 .await;
         }
     }
 }
 
+/// Identifies the data subscriptions that can be handled now.
+/// Returns the list of peers that made those subscriptions
+/// alongside the ledger info at the target version for the peer.
+fn get_peers_with_ready_subscriptions<T: StorageReaderInterface>(
+    cached_storage_server_summary: Arc<RwLock<StorageServerSummary>>,
+    data_subscriptions: Arc<Mutex<HashMap<AccountAddress, DataSubscriptionRequest>>>,
+    storage: T,
+) -> Result<Vec<(AccountAddress, LedgerInfoWithSignatures)>, Error> {
+    // Fetch the latest storage summary and highest synced version
+    let latest_storage_summary = cached_storage_server_summary.read().clone();
+    let highest_synced_ledger_info = match latest_storage_summary.data_summary.synced_ledger_info {
+        Some(ledger_info) => ledger_info,
+        None => return Ok(vec![]),
+    };
+    let highest_synced_version = highest_synced_ledger_info.ledger_info().version();
+    let highest_synced_epoch = highest_synced_ledger_info.ledger_info().epoch();
+
+    // Identify the peers with ready subscriptions
+    let mut ready_subscriptions = vec![];
+    for (peer, data_subscription) in data_subscriptions.lock().iter() {
+        if data_subscription.highest_known_version() < highest_synced_version {
+            let highest_known_epoch = data_subscription.highest_known_epoch();
+            let target_ledger_info = if highest_known_epoch < highest_synced_epoch {
+                // The peer needs to sync to their epoch ending ledger info
+                get_epoch_ending_ledger_info(highest_known_epoch, storage.clone())?
+            } else {
+                highest_synced_ledger_info.clone()
+            };
+            ready_subscriptions.push((*peer, target_ledger_info));
+        }
+    }
+    Ok(ready_subscriptions)
+}
+
+/// Gets the epoch ending ledger info at the given epoch
+fn get_epoch_ending_ledger_info<T: StorageReaderInterface>(
+    epoch: u64,
+    storage: T,
+) -> Result<LedgerInfoWithSignatures, Error> {
+    match storage.get_epoch_ending_ledger_infos(epoch, epoch) {
+        Ok(epoch_change_proof) => {
+            if let Some(ledger_info) = epoch_change_proof.ledger_info_with_sigs.get(0) {
```

Comment:
nit: I think you can avoid `if..else` and do it as one liner by `epoch_change_proof.ledger_info_with_sigs.get(0).ok_or_else(...`